### ISSUE TITLE
xmonad multimonitor support broken

### ISSUE THREAD

**Issue by ericsagnes** [NixOS/nixpkgs#9372]
Xmonad treats all plugged monitors as one big shared workspace. (expected behavior is one workspace per screen)

Note, it was working as expected before I ran a `nixos-rebuild switch --upgrade` today that updated many packages.

xrandr output:
````
Screen 0: minimum 8 x 8, current 5760 x 1080, maximum 32767 x 32767
eDP1 connected 1920x1080+0+0 (normal left inverted right x axis y axis) 344mm x 193mm
   1920x1080     60.06*+
   1400x1050     59.98  
   1600x900      60.00  
   1280x1024     60.02  
   1280x960      60.00  
   1368x768      60.00  
   1280x720      60.00  
   1024x768      60.00  
   1024x576      60.00  
   960x540       60.00  
   800x600       60.32    56.25  
   864x486       60.00  
   640x480       59.94  
   720x405       60.00  
   640x360       60.00  
DP1 disconnected (normal left inverted right x axis y axis)
DP2 disconnected (normal left inverted right x axis y axis)
DP2-1 connected 1920x1080+1920+0 (normal left inverted right x axis y axis) 528mm x 297mm
   1920x1080     60.00*+  50.00    59.94  
   1680x1050     59.88  
   1600x900      59.98  
   1280x1024     60.02  
   1280x800      59.91  
   1280x720      59.97    60.00    50.00    59.94  
   1024x768      60.00  
   800x600       60.32  
   720x576       50.00  
   720x480       60.00    59.94  
   640x480       60.00    59.94  
   720x400       70.08  
DP2-2 connected 1920x1080+3840+0 (normal left inverted right x axis y axis) 509mm x 286mm
   1920x1080     60.00*+  59.94  
   1680x1050     59.88  
   1280x1024     75.02    60.02  
   1440x900      74.98    59.90  
   1280x960      60.00  
   1152x864      75.00  
   1280x720      60.00    50.00    59.94  
   1440x576      50.00  
   1024x768      75.08    70.07    60.00  
   832x624       74.55  
   800x600       72.19    75.00    60.32    56.25  
   720x576       50.00  
   720x480       60.00    59.94  
   640x480       75.00    72.81    66.67    60.00    59.94  
   720x400       70.08  
DP2-3 disconnected (normal left inverted right x axis y axis)
HDMI1 disconnected (normal left inverted right x axis y axis)
HDMI2 disconnected (normal left inverted right x axis y axis)
VGA1 disconnected (normal left inverted right x axis y axis)
VIRTUAL1 disconnected (normal left inverted right x axis y axis)
````
Switching back to an older generation at boot doesn't remove the problem.

nix-channel: nixos https://nixos.org/channels/nixos-unstable

**Comment by vcunat** [comment]
> Switching back to an older generation at boot doesn't remove the problem.

That's the most suspicious point to me.

**Comment by ericsagnes** [comment]
Haskell wiki has some [suggestions](https://wiki.haskell.org/Xmonad/Frequently_asked_questions#xmonad_does_not_detect_my_multi-head_setup) for debugging multi monitors trouble, so I tried to run the command using a nix-shell with the version of ghc that I suppose xmonad is using:

````
nix-shell -p haskellngPackages.ghc --pure
> ghc -e Graphics.X11.Xinerama.compiledWithXinerama
<interactive>:1:1:
    Not in scope: `Graphics.X11.Xinerama.compiledWithXinerama'
````
Note:
xmonad was installed via configuration.nix `xserver.windowManager.xmonad.enable` option.

**Comment by vcunat** [comment]
Any X11 support isn't in the base ghc package, I believe. To see how Haskell stuff works with Nix, best read http://nixos.org/nixpkgs/manual/#how-to-install-a-compiler-with-libraries, I think.

**Comment by ericsagnes** [comment]
Thanks for the advice!
As `Graphics.X11.Xinerama` is in X11-extras, I tried the following:

````
nix-shell -p "haskellngPackages.ghcWithPackages (pkgs: [pkgs.X11 pkgs.X11-extras])"
````

But it failed with the following error:

````
...
Building X11-extras-0.4...
Preprocessing library X11-extras-0.4...

Graphics/X11/Xlib/Extras.hsc:1:12: Warning:
    -fglasgow-exts is deprecated: Use individual extensions instead
[1 of 2] Compiling Graphics.X11.Xlib.Extras ( dist/build/Graphics/X11/Xlib/Extras.hs, dist/build/Graphics/X11/Xlib/Extras.o )

Graphics/X11/Xlib/Extras.hsc:812:25:
    Not in scope: ‘unsafePerformIO’

Graphics/X11/Xlib/Extras.hsc:839:36:
    Not in scope: ‘unsafePerformIO’
builder for ‘/nix/store/4g78gq15crdq7wn5bpvypbnhkpmj9ng4-haskell-X11-extras-0.4.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/ccqd7gk85k3zh2gxq7z89v98w6hsydf9-ghc-7.10.2.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/ccqd7gk85k3zh2gxq7z89v98w6hsydf9-ghc-7.10.2.drv’ failed
/run/current-system/sw/bin/nix-shell: failed to build all dependencies
````

That seems odd as `unsafePerformIO` should be available in base.

**Comment by peti** [comment]
This issue has been fixed in 574e2a5f81f1d13b74c5c1a2c965eef39e4e824d. It's going to take a day or two until that commit shows up in the unstable channel.

**Comment by ericsagnes** [comment]
Thanks!
